crates/sparrow-compiler/src/types/inference.rs

```diff
@@ -109,18 +119,39 @@ pub(crate) fn instantiate(
         }
     }
 
-    let solutions: HashMap<TypeConstraint, FenlType> = types_for_constraints
+    let solutions: HashMap<TypeVariable, FenlType> = types_for_variable
```

Comment:
Rather than running find over the parameters every time, I think you could iterate over the parameters and pair that with the type variables.

```
let solutions = signature.type_parameters.iter().map(|p| {
  // This expect should never fail -- it would mean no arguments used the type variable
  // Which would have been an invalid signature.
  let argument_types = types_for_variable.remove(p.name).expect("unused type variable");
  ...
}).try_collect()?;

// Then check to see if types_for_variable is empty -- if not, then there is a different error.
// Not sure it needs a diagnostic though since it means we had a type variable in the
// signature that wasn't defined.
```